pyoxidizer/src/py_packaging/distribution.rs

```diff
@@ -360,24 +475,170 @@ impl ParsedPythonDistribution {
+                );
+            };
+        }
+
+        let python_paths = resolve_python_paths(&venv_base, &self.version);
+
+        invoke_python(&python_paths, &logger, &["-m", "ensurepip"]);
+
+        if VENV_MODULE == "virtualenv" {
+            invoke_python(&python_paths, &logger, &["-m", "pip", "install", "virtualenv"]);
+        }
+
+        prepare_hacked_distutils(logger, &python_paths);
+
+        python_paths
+    }
+
+    /// Create a venv from the distribution at path.
+    pub fn create_venv(&self, logger: &slog::Logger, path: &PathBuf) -> PythonPaths {
+        let venv_dir_s = path.display().to_string();
+
+        // This will recreate it, if it was deleted
+        let python_paths = self.create_hacked_base(&logger);
+
+        if path.exists() {
+            warn!(logger, "re-using {} {}", VENV_MODULE, venv_dir_s);
+        } else {
+            warn!(logger, "creating {} {}", VENV_MODULE, venv_dir_s);
+            invoke_python(&python_paths, &logger, &["-m", VENV_MODULE, venv_dir_s.as_str()]);
+        }
+
+        resolve_python_paths(&path, &self.version)
+    }
+
+    /// Create or re-use an existing venv
+    pub fn prepare_venv(
+        &self,
+        logger: &slog::Logger,
+        venv_path: Option<&String>,
+    ) -> (PythonPaths, HashMap<String, String>) {
+        let venv_dir_path = match venv_path {
+            Some(path_str) => PathBuf::from(path_str),
+            None => tempdir::TempDir::new("pyoxidizer-temp-venv")
+                .expect("could not create temp directory")
+                .path()
+                .join("venv"),
+        };
+
+        let python_paths = self.create_venv(logger, &venv_dir_path);
+
+        let mut extra_envs = HashMap::new();
+
+        let prefix_s = python_paths
+            .prefix
+            .canonicalize()
+            .unwrap()
+            .display()
+            .to_string();
+
+        let venv_path_bin_s = python_paths
+            .bin_dir
+            .canonicalize()
+            .unwrap()
+            .display()
+            .to_string();
+
+        let path_separator = if cfg!(windows) { ";" } else { ":" };
+
+        let process_path_s = env::var("PATH").unwrap();
+
+        extra_envs.insert(
+            "PATH".to_string(),
+            format!("{}{}{}", venv_path_bin_s, path_separator, process_path_s),
+        );
+
+        let mut site_packages_s = python_paths
+            .site_packages
+            .canonicalize()
+            .unwrap()
+            .display()
+            .to_string();
+        if site_packages_s.starts_with("\\\\?\\") {
+            site_packages_s = site_packages_s[4..].to_string();
+        }
+
+        extra_envs.insert("VIRTUAL_ENV".to_string(), prefix_s);
+        extra_envs.insert("PYTHONPATH".to_string(), site_packages_s);
+
+        fs::create_dir_all(&python_paths.pyoxidizer_state_dir).unwrap();
+
+        extra_envs.insert(
+            "PYOXIDIZER_DISTUTILS_STATE_DIR".to_string(),
```

Comment:
this is moved out of distutils.rs because most of the venv environ stuff is here, and it feels a bit odd to getting only a single path returned from `prepare_hacked_distutils` and that value does not need to be any specific value for the distutils hack to work.

Reply:
Also being able to run commands in the venv outside of PyOxidizer provides a lot of flexibility not currently possible with the config file.  For that to happen either:

1. the distutils hacking pushes the state directory name into the .py files, or
2. the distutils .py and rust code both use the same state directory path

I have opt for the latter, as that seems to be more appropriate in a venv which has a 'prefix'